File: cpp/Remove_Element.cc
/*
Given an array and a value, remove all instances of that value in place and return the new length.

The order of elements can be changed. It doesn't matter what you leave beyond the new length.
*/

class Solution 
{
public:
    int removeElement(int A[], int n, int elem) 
    {
        int num = 0;
        for(int i = 0; i < n-num; i++)
        {
            if (A[i] == elem)
            {
                A[i] = A[n-1-num]; // 把最后的元素放到这个'坑'
                num++;
                i--;
            }
        }
        return n-num;
    }
    /*更简单点的
    int removeElement(int A[], int n, int elem) 
    {
        int index = 0;
        for (int i = 0; i < n; ++i) 
        {
            if (A[i] != elem) 
            {
                A[index++] = A[i];
            }
        }
        return index;
    }
    */
};

File: go/remove-element/remove-element.go
package remove_element

/*
Given an array and a value, remove all instances of that value in-place and return the new length.

Do not allocate extra space for another array,
you must do this by modifying the input array in-place with O(1) extra memory.

The order of elements can be changed. It doesn't matter what you leave beyond the new length.

Example:

Given nums = [3,2,2,3], val = 3,

Your function should return length = 2, with the first two elements of nums being 2.
*/


func removeElement(nums []int, val int) int {
    index := 0

    for _, n := range nums{
        if n != val {
            nums[index] = n
            index++
        }
    }

    return index
}